src/apps/chifra/pkg/cache/read.go

```diff
@@ -0,0 +1,813 @@
+	if err != nil {
+		return
+	}
+	return
+}
+
+func readAddress(reader *bufio.Reader, target *common.Address) (err error) {
+	str := &cString{}
+	err = readCString(reader, str)
+	if err != nil {
+		return
+	}
+	addr := common.HexToAddress(string(str.content))
+	*target = addr
+	return
+}
+
+func readHash(reader *bufio.Reader, target *common.Hash) (err error) {
+	str := &cString{}
+	err = readCString(reader, str)
+	if err != nil {
+		return
+	}
+	hash := common.HexToHash(string(str.content))
+	*target = hash
+	return
+}
+
+func readString(reader *bufio.Reader, target *string) (err error) {
+	str := &cString{}
+	err = readCString(reader, str)
+	if err != nil {
+		return
+	}
+	s := string(str.content)
+	*target = s
+	return
+}
+
+// readBigUint reads biguint from cache file and converts it to Go's big.Int
+func readBigUint(reader *bufio.Reader, target *big.Int) (err error) {
+	read := createReadFn(reader)
+	// read capacity
+	var capacity int32
+	err = read(&capacity)
+	if err != nil {
+		return
+	}
+	// read length
+	var length int32
+	err = read(&length)
+	if err != nil {
+		return
+	}
+	if length == 0 {
+		*target = *big.NewInt(0)
+		return
+	}
+
+	// read items as slice of bytes
+	items := make([]byte, length*8)
+	err = read(&items)
+	if err != nil {
+		return
+	}
+	result := big.NewInt(0)
+	// in order to get the correct number, we have to manually reverse bytes
+	// see https://github.com/golang/go/issues/40891
+	orderedItems := reverseBytes(items)
+	// load bytes into big.Int
+	result.SetBytes(orderedItems)
+	*target = *result
+	return
+}
+
+// makeArrayItemRead takes a function that reads bytes and changes variable's value
+// and turns it into a function returning a value. Such a function can then be provided
+// to `readFromArray` to read and convert array items
+func makeArrayItemRead[Item ArrayItem](
+	reader *bufio.Reader,
+	readAndSetValue func(reader *bufio.Reader, target *Item) (err error),
+) func(reader *bufio.Reader) (value *Item, err error) {
+	return func(reader *bufio.Reader) (value *Item, err error) {
+		v := new(Item)
+		err = readAndSetValue(reader, v)
+		value = v
+		return
+	}
+}
+
+func readTimestamp(read readBytes, target *time.Time) (err error) {
+	var rawTimestamp int64
+	err = read(&rawTimestamp)
+	if err != nil {
+		return
+	}
+	*target = time.Unix(rawTimestamp, 0)
+	return
+}
+
+func readUintAsBool(read readBytes, target *bool) (err error) {
+	var raw uint8
+	err = read(&raw)
+	if err != nil {
+		return
+	}
+	*target = raw != 0
+	return
+}
+
+func readCacheHeader(reader *bufio.Reader, target *cacheHeader) (err error) {
+	read := createReadFn(reader)
+
+	err = read(&target.deleted)
+	if err != nil {
+		return
+	}
+
+	err = read(&target.schema)
+	if err != nil {
+		return
+	}
+
+	err = read(&target.showing)
+	if err != nil {
+		return
+	}
+
+	err = readCString(reader, &target.className)
+	if err != nil {
+		return err
+	}
+
+	return
+}
+
+func ReadBlock(reader *bufio.Reader) (block *types.SimpleBlock, err error) {
```

Comment:
I feel like eventually we will want to move this data type specific code to a data type specific file. In this way, it can be more easily auto-generated. Don't change this, but please add a TODO list item saying something similar to the note I asked you to add above (So I can search for and find both with one search.)

Same for ReadTransaction below.